artiq/test/coredevice/test_exceptions.py

```diff
@@ -0,0 +1,92 @@
+import re
+
+from artiq.experiment import *
+from artiq.master.worker_db import DeviceError
+from artiq.test.hardware_testbench import ExperimentCase
+
+
+class CustomException(Exception):
+    pass
+
+
+class KernelFmtException(EnvExperiment):
+    def build(self):
+        self.setattr_device("core")
+
+    @kernel
+    def run(self):
+        self.throw()
+
+    def throw(self):
+        raise CustomException("{foo}")
+
+
+class KernelNestedFmtException(EnvExperiment):
+    def build(self):
+        self.setattr_device("core")
+
+    @kernel
+    def run(self):
+        try:
+            self.throw_foo()
+        except:
+            try:
+                raise RTIOUnderflow("{bar}")
+            except:
+                try:
+                    raise RTIOOverflow("{bizz}")
+                except:
+                    self.throw_buzz()
+
+    def throw_foo(self):
+        raise CustomException("{foo}")
+
+    def throw_buzz(self):
+        raise RTIOUnderflow("{buzz}")
+
+
+class KernelRTIOUnderflow(EnvExperiment):
+    def build(self):
+        self.setattr_device("core")
+        try:
+            self.setattr_device("led")
+        except DeviceError:
+            self.led = self.get_device("led0")
+
+    @kernel
+    def run(self):
+        self.core.reset()
+        for _ in range(1000):
+            self.led.on()
+            self.led.off()
```

Comment:
This seems like a very roundabout/brittle way of inducing an underflow. Perhaps just `at_mu(self.core.get_rtio_counter_mu() - 1000); self.led.on()` or something like that?

Reply:
I had the same observation, but figured the channel presumably has event replacement enabled, and maybe that is then the correct behavior?

Reply:
I got underflow, as described in the tests

Reply:
Maybe splitting this off into a separate issue and replacing this with a test case that actually is supposed to underflow is the way to go?

Reply:
In the docs:
> If the current cursor is in the past, an [artiq.coredevice.exceptions.RTIOUnderflow](https://m-labs.hk/artiq/manual/core_drivers_reference.html#artiq.coredevice.exceptions.RTIOUnderflow) exception is thrown.

Isn't it actually this at `at_mu(self.core.get_rtio_counter_mu() - 1000); self.led.on()` ? So it seems to be fine here